Sources/Shared/LocalAuthentication/BiometricsAuthenticationService.swift

```diff
@@ -0,0 +1,142 @@
+                completion?()
+            }
+        } else {
+            removeProtectionOverlay()
+        }
+    }
+
+    #if targetEnvironment(macCatalyst)
+    /// This avoids macOS going on loop between enter background and enter foreground while trying to authenticate
+    @objc private func allowReauthenticate() {
+        shouldAllowMacToAuthenticate = true
+    }
+    #endif
+
+    private func authenticate() {
+        var error: NSError?
+        context?.invalidate()
+        context = LAContext()
+        if context?.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) ?? false {
+            authenticate(policy: .deviceOwnerAuthenticationWithBiometrics)
+        } else if context?.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) ?? false {
+            authenticate(policy: .deviceOwnerAuthentication)
+        } else {
+            Current.Log
+                .error(
+                    "Failed to authenticate with biometrics, user context can't evaluate policy for device ownership"
+                )
+        }
+    }
+
+    private func removeProtectionOverlay() {
+        DispatchQueue.main.async { [weak self] in
+            self?.overlayViewController?.dismiss(animated: true) {
+                self?.overlayViewController = nil
+            }
+        }
+    }
+
+    private func authenticate(policy: LAPolicy) {
+        context?
+            .evaluatePolicy(
+                policy,
+                localizedReason: L10n.SettingsDetails.General.Security
+                    .authenticationPolicy
+            ) { [weak self] authorized, error in
+                self?.didFinishAuthentication(authorized: authorized)
+
+                if let error {
+                    Current.Log.error(["Failed to evaluate policy for authentication": error.localizedDescription])
+                }
+            }
+    }
+
+    private func didFinishAuthentication(authorized: Bool) {
+        if authorized {
+            #if targetEnvironment(macCatalyst)
+            shouldAllowMacToAuthenticate = false
+            #endif
+            context?.invalidate()
+            context = nil
+            removeProtectionOverlay()
+        } else {
+            overlayViewController?.updateUnlockButtonVisibility(visible: true)
+        }
+    }
+
+    private func addBiometricOverlayProtection(controller: UIViewController, completion: @escaping () -> Void) {
+        guard overlayViewController == nil else {
+            completion()
+            return
+        }
+        overlayViewController = BiometricsAuthenticationViewController()
+        overlayViewController?.modalPresentationStyle = .overCurrentContext
+        overlayViewController?.delegate = self
+        guard let overlayViewController = overlayViewController else { return }
+
+        DispatchQueue.main.async {
+            controller.present(overlayViewController, animated: false) {
```

Comment:
what happens if a modal is already being presented?

i think we may need to do something more like:

1. insert this into the window at the top
2. do the appearance manually (beginTransition/endTransition)
3. make sure it stays on top